replay2/memory_remapper/tests/memory_remapper_tests.cc

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2022 Google Inc.
+//
+// This file is generated code. It was created by the AGI code generator.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include "replay2/memory_remapper/memory_remapper.h"
+
+using namespace agi::replay2;
+
+class ModResourceGenerator : public ResourceGenerator {
+public:
+	ModResourceGenerator(size_t length) : length_(length) {}
+	virtual ~ModResourceGenerator() {}
+
+	size_t length() const override { return length_; }
+
+	void generate(ReplayAddress replayAddress) override {
+		for(size_t i = 0; i < length_; ++i) {
+			replayAddress.bytePtr()[i] = std::byte(i % 256);
+		}
+	}
+private:
+	size_t length_;
+};
+
+class ConstResourceGenerator : public ResourceGenerator {
+public:
+	ConstResourceGenerator(std::byte value, size_t length) : value_(value), length_(length) {}
+	virtual ~ConstResourceGenerator() {}
+
+	size_t length() const override { return length_; }
+
+	void generate(ReplayAddress replayAddress) override {
+		for(size_t i = 0; i < length_; ++i) {
+			replayAddress.bytePtr()[i] = value_;
+		}
+	}
+private:
+	std::byte value_;
+	size_t length_;
+};
+
+void ASSERT_MOD_REPLAY_ADDRESS(const MemoryRemapper& remapper, const CaptureAddress& captureAddress, const ReplayAddress& replayAddress, size_t length) {
+	for(size_t i = 0; i < length; ++i) {
+		ReplayAddress replayAddress = remapper.RemapCaptureAddress(captureAddress.offsetByBytes(i));
+		ASSERT_EQ(replayAddress.bytePtr()[0], std::byte(i % 256));
+	}
+}
+
+void ASSERT_CONST_REPLAY_ADDRESS(const MemoryRemapper& remapper, const CaptureAddress& captureAddress, const ReplayAddress& replayAddress, std::byte value, size_t length) {
+	for(size_t i = 0; i < length; ++i) {
+		ReplayAddress replayAddress = remapper.RemapCaptureAddress(captureAddress.offsetByBytes(i));
+		ASSERT_EQ(replayAddress.bytePtr()[0], value);
+	}
+}
+
+TEST(MemoryRemapperTests, SimpleMapping) {
+
+	const size_t size = 128;
+
+	std::byte* rawCapturePtr = new std::byte[size];
+	CaptureAddress captureAddress(rawCapturePtr);
+
+	MemoryRemapper remapper;
+	const MemoryObservation captureObservation(captureAddress, std::make_shared<ModResourceGenerator>(size));
```

Comment:
What will the usage pattern will look like for real resource generators with replaying? This doesn't look safe. We should be working with ranges to make sure the size in the resource generator does not override the allocated block. Actually, if that's applied, 'size' in resource generator will look redundant with current generators.

Reply:
A memory observation will come from a capture file. The interceptor saw X bytes of memory containing data Y at address Z. So we will have something like:

const MemoryObservation captureObservation(Z, std::make_shared<RealResourceGenerator>(Y, X));

Where X, Y and Z all come out of fread() into the file.

Reply:
If our future transform system was going to add some commands to a trace, and those commands required some memory observations to be constructed, we'd need to find some new address ranges that are not already used by other commands in the trace. This would require some sort of custom address allocator that would take a size S and return you a new CaptureAddressRange such that [S, S+n) didn't overlap with anything in the capture. We don't need to support this usecase yet so I haven't included such an allocator here.